src/downloadlist.cpp

```diff
@@ -35,33 +35,36 @@ DownloadList::DownloadList(OrganizerCore& core, QObject* parent)
     : QAbstractTableModel(parent), m_manager(*core.downloadManager()),
       m_settings(core.settings())
 {
-  connect(&m_manager, SIGNAL(update(int)), this, SLOT(update(int)));
-  connect(&m_manager, SIGNAL(aboutToUpdate()), this, SLOT(aboutToUpdate()));
+  connect(&m_manager, SIGNAL(update(DownloadManager::DownloadInfo*)), this,
+          SLOT(update(DownloadManager::DownloadInfo*)));
+  connect(&m_manager, SIGNAL(downloadAdded(DownloadManager::DownloadInfo*)), this,
+          SLOT(downloadAdded(DownloadManager::DownloadInfo*)));
+  connect(&m_manager, SIGNAL(downloadRemoved(QUuid)), this,
+          SLOT(downloadRemoved(QUuid)));
+  connect(&m_manager, SIGNAL(pendingDownloadAdded(QString)), this,
+          SLOT(pendingDownloadAdded(QString)));
+  connect(&m_manager, SIGNAL(pendingDownloadRemoved(QString)), this,
+          SLOT(pendingDownloadRemoved(QString)));
 }
 
-int DownloadList::rowCount(const QModelIndex& parent) const
+QModelIndex DownloadList::index(int row, int column, const QModelIndex&) const
 {
-  if (!parent.isValid()) {
-    // root item
-    return m_manager.numTotalDownloads() + m_manager.numPendingDownloads();
-  } else {
-    return 0;
-  }
+  return createIndex(row, column, row);
```

Comment:
I'm not sure that passing the row as the id of createIndex() is the correct solution. It sounds like we would end up relying on the row to be consistent, which sounds a bit too much like the problem that we are trying to solve already. 

From the docs it looks like you can pass a pointer in here, but honestly I'm not sure whether that is what we are supposed to be using.

I'm hoping someone can chime in and clear things up here.

Reply:
Yes, this is pretty dangerous... Means you need to recompute everything every time something changes in the list of downloads.